src/main/java/programminglife/gui/controller/GraphController.java

```diff
@@ -0,0 +1,173 @@
+     */
+    private Set<Node> drawDFS(Node origin, Node node, XYCoordinate offset) {
+        return this.drawDFS(origin, node, offset, -1, new HashSet<>());
+    }
+
+    /**
+     * Draw all nodes recursively on the screen.
+     *
+     * @param origin The parent {@link Node} that initiated this draw call
+     * @param node Draw this {@link Node} and all its children recursively
+     * @param offset Draws nodes at this offset from the top-left of the screen
+     * @param maxDepth The max depth from root to draw nodes
+     * @return a {@link Set} of all drawn {@link Node}s
+     */
+    public Set<Node> drawDFS(Node origin, Node node, XYCoordinate offset, int maxDepth) {
+        return this.drawDFS(origin, node, offset, maxDepth, new HashSet<>());
+    }
+
+    /**
+     * Draw all nodes recursively on the screen.
+     * @param origin
+     * @param node Draw this node and all its children recursively
+     * @param offset Draws nodes at this offset from the top-left of the screen
+     * @param drawnNodes A set containing all drawn nodes
+     */
+    private Set<Node> drawDFS(Node origin, Node node, XYCoordinate offset, int maxDepth, Set<Node> drawnNodes) {
+        boolean nodeIsDrawn = drawnNodes.contains(node);
+        if (!nodeIsDrawn) {
+            node.setLocation(offset);
+        }
+
+        if (origin != null && maxDepth != 0) {
+            XYCoordinate targetLeft = node.getLeftBorderCenter();
+            XYCoordinate originRight = origin.getRightBorderCenter();
+
+            Line link = new Line(targetLeft.getX(), targetLeft.getY(), originRight.getX(), originRight.getY());
+            link.setStroke(Color.DARKGRAY);
+            link.setStrokeWidth(3);
+            link.setOnMouseClicked(event -> this.console.appendText(String.format("Link{%s -> %s}\n", origin, node)));
+
+            this.grpDrawArea.getChildren().add(link);
+        }
+
+        if (maxDepth != 0 && !nodeIsDrawn) {
+            this.drawNode(node);
+            drawnNodes.add(node);
+
+            int childCount = 0;
+            for (Node child : node.getChildren()) {
+                XYCoordinate newOffset = offset.add(HORIZONTAL_OFFSET)
+                        .add(node.getWidthCoordinate())
+                        .setY(INITIAL_OFFSET.getY() + (int) (CHILD_OFFSET * childCount * node.getHeight()));
+                this.drawDFS(node, child, newOffset, maxDepth - 1, drawnNodes);
+                childCount++;
+            }
+        }
+
+        return drawnNodes;
+    }
+
+    /**
+     * Draws the node on the canvas.
+     *
+     * @param nodeID The ID of the node to draw
+     * @return The size of the node
+     */
+    private XYCoordinate drawNode(int nodeID) {
+        return this.drawNode(this.graph.getNode(nodeID));
+    }
+
+    /**
+     * Draw a single {@link Node}.
+     * @param node the {@link Node} to draw
+     * @return the size of the drawn {@link Node}
+     */
+    private XYCoordinate drawNode(Node node) {
```

Comment:
For the next sprint, the dumb drawing should probably happen in a seperate class in a seperate package from the controller.